Stage 3 Review

MenasheKoren opened the pull request.

MenasheKoren committed:
Feat: WIP add NewsApi.js

MenasheKoren committed:
Feat: WIP add api content to newscards

MenasheKoren committed:
Feat: WIP Change NewsApi to class

MenasheKoren committed:
Feat: Add NewsApi and add functionality to show more button

MenasheKoren committed:
Feat: WIP Add functionality to search bar

MenasheKoren committed:
Feat: Add functions to search and show more buttons

MenasheKoren committed:
Hotfix: Minor fix

MenasheKoren committed:
Hotfix: Update NothingFound logic

MenasheKoren committed:
Hotfix: Update NothingFound logic

MenasheKoren committed:
Delete .env

MenasheKoren committed:
Feat: WIP add context and auth files

MenasheKoren committed:
Feat: WIP Update Register.js

MenasheKoren committed:
Feat: Add FormContext.js

MenasheKoren committed:
Hotfix: WIP fix Register.js

MenasheKoren committed:
Hotfix: WIP fix Register.js

MenasheKoren committed:
Merge remote-tracking branch 'origin/feature/register' into feature/register

MenasheKoren committed:
Hotfix: Enable Register.js

MenasheKoren committed:
Feat: WIP Add AppStateContext.js

MenasheKoren committed:
Hotfix: Remove AuthStateContext.js

MenasheKoren committed:
Merge remote-tracking branch 'origin/feature/register' into feature/register

# Conflicts:
#	src/components/App/App.js
#	src/components/HeaderButton/HeaderButton.js

MenasheKoren committed:
Feat: WIP Add saved news apis

MenasheKoren committed:
Feat: Update InfoToolTip.js

MenasheKoren committed:
Feat: Update authentication

MenasheKoren committed:
Feat: WIP update bookmark functioning

MenasheKoren committed:
Feat: WIP update bookmark functioning

MenasheKoren committed:
Merge remote-tracking branch 'origin/feature/saved-news' into feature/saved-news

# Conflicts:
#	src/components/NewsCard/NewsCard.js

MenasheKoren committed:
Feat: WIP update currentUser on login

MenasheKoren committed:
Hotfix: WIP move handleLogin to App.js

MenasheKoren committed:
Feat: Enable Login.js

MenasheKoren committed:
Hotfix: Minor fix

MenasheKoren committed:
Feat: Enable saving articles

MenasheKoren committed:
Feat: WIP Add saved article cards

MenasheKoren committed:
Feat: WIP set saved cards atts

MenasheKoren committed:
Feat: Add atts to saved articles

MenasheKoren committed:
Feat: Add logic for keyword list

MenasheKoren committed:
Feat: WIP add refactored files and remove unused

MenasheKoren committed:
Feat: Add keyword list

MenasheKoren committed:
Merge pull request #21 from MenasheKoren/feature/keyword-list

Complete Stage 3

MenasheKoren committed:
Hotfix: Minor fixes

gennady-bars commented on src/utils/MainApi.js:
```diff
@@ -0,0 +1,50 @@
+
+const customFetch = (url, headers) =>
+  fetch(url, headers).then((res) =>
+    res.ok ? res.json() : Promise.reject(res.statusText)
```
you need to use `checkResponse` for checking responses not to duplicate the code

gennady-bars commented on src/utils/NewsApi.js:
```diff
@@ -0,0 +1,43 @@
+    return fetch(
+      `${this.baseUrl}q=${query}&language=en&from=${this.date.from}&to=${this.date.to}&pageSize=100&apiKey=${this.apiKey}`
+    )
+      .then((res) => {
```
* you need to use `checkResponse` for checking responses not to duplicate the code

gennady-bars commented on src/utils/NewsApi.js:
```diff
@@ -0,0 +1,43 @@
+          return { title, source, date, text, link, image };
+        });
+      })
+      .catch((err) => {
```
It’s better not to use `catch` in `NewsApi` (or any Api) because you need to `catch` possible errors at the very end of any request and this end is in `index.js` (`App.js`)  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

gennady-bars commented on src/utils/auth.js:
```diff
@@ -0,0 +1,57 @@
+    });
+};
+
+export const getContent = () => {
```
`getContent` should be named `checkToken`, because you check a token here

gennady-bars commented on src/components/App/App.js:
```diff
@@ -3,29 +3,61 @@ import { useEffect, useState } from "react";
+
+  useEffect(() => {
+    if (localStorage.getItem("token") !== null) {
+      localStorage.getItem("token");
```
`localStorage.getItem("token");`

this has no effect.

you need to pass the toke to the call of `getContent` 


```js


    const token = localStorage.getItem("token");
    if (token) {
      auth
        .getContent(token)

```

gennady-bars commented on src/components/App/App.js:
```diff
@@ -3,29 +3,61 @@ import { useEffect, useState } from "react";
+          setIsLoggedIn(true);
+          setCurrentUser(user);
+        })
+        .then(
```
In blocks `then`, `catch`, `finally` you need to pass only the reference to the function (not a function call) otherwise the function is called immediately without waiting for the response from the server.


```js

        .then(() => {
          mainApi
            .getSavedArticles()
            .then((data) => {
              setSavedCards(data);
            })
            .catch((err) => console.log(`Error..... ${err}`));
        })
        .catch((err) => console.log(`Error..... ${err}`));

```

gennady-bars commented on src/components/FormInput/FormInput.js:
```diff
@@ -0,0 +1,34 @@
+        placeholder={placeholder}
+        id={id}
+        name={name}
+        defaultValue={defaultValue}
```
`defaultValue` is used only with `useRef`. For controlled components you need to use only `value`

gennady-bars commented on src/components/PopupWithForm/PopupWithForm.js:
```diff
@@ -6,10 +6,14 @@ import { Login } from "../Login/Login";
 
 export function PopupWithForm({
```
please, make `PopupWithForm` the same way you had in the `React` project: it's name says: it is a popup with a form inside it. So, you need to have tag `form`, `submit button`, cross button inside `PopupWithForm`. And any content of this compment should be passed as its children.


It should be a universal compoent for any popups with a form. Not just for Register or Login

gennady-bars commented on src/components/Register/Register.js:
```diff
@@ -1,26 +1,44 @@
+  }
+
+  useEffect(() => {
     const validatedForm = new FormValidator(formSettings, formRef.current);
```
**Might be better**

that's a `js`-variant of validation. 

If it’s interesting for you here is a custom `validation hook`.

You should put it in a separate file  `useFormAndValidation.js` in folder `hooks` and then import it to the component that needs validation:

```js

import {useState, useCallback} from 'react';

export function useFormAndValidation() {
  const [ values, setValues ] = useState({});
  const [ errors, setErrors ] = useState({});
  const [ isValid, setIsValid ] = useState(true);

  const handleChange = (e) => {
    const {name, value} = e.target
    setValues({...values, [name]: value });
    setErrors({...errors, [name]: e.target.validationMessage});
    setIsValid(e.target.closest('form').checkValidity());
  };



  const resetForm = useCallback((newValues = {}, newErrors = {}, newIsValid = false) => {
    setValues(newValues);
    setErrors(newErrors);
    setIsValid(newIsValid);
  }, [setValues, setErrors, setIsValid]);

  return { values, handleChange, errors, isValid, resetForm, setValues, setIsValid };
}


```

And only one line of code enables validation:

```js

  const {values, handleChange, errors, isValid, setValues, resetForm} = useFormAndValidation()

```

So, you don’t need to make functions for handling inputs, storing values, resetting the form. Everything is already done.

gennady-bars reviewed (commented):
Dear Menashe , (**Please, expand the general comment ↓**)

Good job but there are some remarks:


* In ` README.md` you need to add some info about your project: technologies, how to download, how to start, how to deploy the app. And you need to add the link to your deployed app.
* You need to deploy the app
* Plaese, delete 3 CRA demo files. See here http://joxi.ru/bmoKL1VuO370Ym 
* you need to use `checkResponse` for checking responses not to duplicate the code
* It’s better not to use `catch` in `NewsApi` (or any Api) because you need to `catch` possible errors at the very end of any request and this end is in `index.js` (`App.js`)  after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* `getContent` should be named `checkToken`, because you check a token here
* In blocks `then`, `catch`, `finally` you need to pass only the reference to the function (not a function call) otherwise the function is called immediately without waiting for the response from the server.
* `defaultValue` is used only with `useRef`. For controlled components you need to use only `value`
* please, make `PopupWithForm` the same way you had in the `React` project: it's name says: it is a popup with a form inside it. So, you need to have tag `form`, `submit button`, cross button inside `PopupWithForm`. And any content of this compment should be passed as its children. It should be a universal compoent for any popups with a form. Not just for Register or Login
* if I have seached some info and after that log in I can't save the articles. See the errors http://joxi.ru/p27YKB7SLKnVBr . Only after reloading the page I can do that.


**Might be better**

  * <!--2.16--> When trying to access the `/saved-news` route, unauthorized users are redirected to `/signin` with an open authorization popup window.
  * <!--2.16--> When an unauthorized user clicks on the icon to save an article, the 
  *  <!--2.16--> The necessary handlers are defined inside the `App` root component: `onRegister()`, `onLogin()`, and `onSignOut()`. These handlers are passed to the appropriate components: `Register.js`, `Login.js`, and `NavBar.js`.
  * <!--1.88--> Arrow functions are not passed as handler functions.
  * <!--1.88--> The user receives a message in case of an error.

Please, correct everything and your project will be accepted.

Good luck with the refactoring



```markdown
# Summary:
### Performance Criteria

- [] <!--30.27--> The project functionality is fully implemented according to the current stage's requirements:
    - The preloader is visible and spinning during request execution.
    - All project links and buttons are functioning.
    - All forms are validated on the client side.
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the `/signup` route, provided that all input fields have been filled in correctly. If the request is successful, a message is displayed.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a log out button, should appear in its place.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the `/articles` route of the Practicum API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the following: the keyword by which the card was found and the trash can icon for deleting the article.
- [x] <!--2.16--> No errors occur when building or running the project.
- [x] <!--2.16--> The global state variable `currentUser` has been created.
- [x] <!--2.16--> Local storage: the interaction with the JWT token, the results of the search request, and the state variable are set up in the right order.
- Registration and authorization:
    - [x] <!--2.16--> The `/saved-news` route is protected using the `ProtectedRoute` HOC component.
    - [] <!--2.16--> When trying to access the `/saved-news` route, unauthorized users are redirected to `/signin` with an open authorization popup window.
    - [] <!--2.16--> When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - [x] <!--2.16--> After successful authorization, the popup window is closed. A link to "Saved articles" and a logout button with the username appear in the header.
    - [x] <!--2.16--> The `/` route is not protected.
    - [x] <!--2.16--> After a successful `onSignOut()` handler call, the user is redirected to `/`.
    - [x] <!--2.16--> The `useHistory()` hook is used correctly.
    - [x] <!--2.16--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- Components:
    - [x] <!--2.16--> Hooks are not used inside conditional statements or loops.
    - [x] <!--2.16--> For class components, effects are described inside the component lifecycle methods.
    - [x] <!--2.16--> Hooks are called in a component's main function.
    - [x] <!--2.16--> The `Main`, `NavBar`, and `SavedNewsHeader` components are subscribed to the `CurrentUserContext` context.
    - [x] <!--2.16--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
    - [x] <!--2.16--> The `currentUser` state variable is created in the `App` root component. This variable is used as a value of the context provider.
    - [] <!--2.16--> The necessary handlers are defined inside the `App` root component: `onRegister()`, `onLogin()`, and `onSignOut()`. These handlers are passed to the appropriate components: `Register.js`, `Login.js`, and `NavBar.js`.
    - [x] <!--2.16--> The popup components only contain the submit event handlers. Other handlers, such as `handleUpdateUser()`, are described inside the `App` component.
- [x] <!--2.16--> Asynchronous API requests:
	- Requests can be made through the Fetch API or by using XMLHttpRequest.
	- Third-party libraries (such as axios or jQuery) are not used.
	- API requests are contained in separate files: `MainApi.js` and `NewsApi.js`.
	- The chain for processing promises ends with a `catch()` block.
	- The first `then()` handler returns `res.json`.
- [x] <!--2.16--> Naming:
	- Function and variable names use camelCase.
	- Only nouns are used as variable names.
	- Plural nouns are used for `NodeList`.
	- Classes are named using nouns that begin with a capital letter.
	- The `const` keyword is used for variables that won't be changed directly.
	- Class names correspond with their content.
	- Names must not include inappropriate or unclear abbreviations.
- [x] <!--2.16--> No third-party JavaScript libraries are used.
- [x] <!--2.16--> The date parameters of requests to News API must be calculated automatically (7 days prior to the current date). We recommend converting the `from` parameter to a timestamp and vice versa.

### Best Practices

- [x] <!--1.88--> The initial state of state variables contain the correct data type.
- [x] <!--1.88--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--1.88--> API requests are described inside the `App` component.
- [] <!--1.88--> Arrow functions are not passed as handler functions.
- [] <!--1.88--> API error handling:
    - [] <!--1.88--> The user receives a message in case of an error.
    - [] <!--1.88--> Form fields are blocked while requests are being sent.
- [x] <!--1.88--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.67--> Semantically correct blocks are used for components.
- [x] <!--1.67--> No `<div>` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.67--> The code is clean and easy to understand:
    - The purpose of each listener callback is clear from its name.
    - There are no "magic numbers," i.e. all numeric values are assigned to variables.
    - Сonstants are described in the same scope in which the function is declared. Inside the function, the values of the constants are taken from the closure.

## Number of points: 56

## Additional comments:
- 
```

MenasheKoren committed:
Hotfix: Fix reviewer comments.

1. Remove demo CRA files.
2. Remove catch from NewsApi.js.

MenasheKoren committed:
Hotfix: Change getContent to checkToken

MenasheKoren committed:
Hotfix: Change defaultValue in forms to value

MenasheKoren committed:
Hotfix: Fix reviewer comments.

1. Use checkResponse.
2. Add token const.
3. WIP Add useFormAndValidation.js.
4. WIP Refactor PopupWithForm.js.
5. WIP Refactor ProtectedRoute.js.

MenasheKoren committed:
Hotfix: WIP Fix forms and bookmarks

MenasheKoren committed:
Hotfix: Add register popup for unauthorized actions

MenasheKoren committed:
Hotfix: WIP Remove useFormAndValidation.js

MenasheKoren committed:
Hotfix: WIP Remove useFormAndValidation.js

MenasheKoren committed:
Merge remote-tracking branch 'origin/hotfix/reviewer-1st' into hotfix/reviewer-1st

# Conflicts:
#	src/components/App/App.js
#	src/components/PopupWithForm/PopupWithForm.js

MenasheKoren committed:
Hotfix: WIP fix InfoToolTip.js

MenasheKoren committed:
Hotfix: WIP Update ProtectedRoute.js

MenasheKoren committed:
Hotfix: Update ProtectedRoute.js

MenasheKoren committed:
Hotfix: Update README.md

MenasheKoren committed:
Hotfix: Update README.md

MenasheKoren committed:
Merge remote-tracking branch 'origin/hotfix/reviewer-1st' into hotfix/reviewer-1st

# Conflicts:
#	README.md

MenasheKoren committed:
Hotfix: Update deploy script

gennady-bars reviewed (approved):
Dear Menashe ,

You’ve done a great job.

Your project has been accepted.

Good luck




```markdown
# Summary:
### Performance Criteria

- [x] <!--30.27--> The project functionality is fully implemented according to the current stage's requirements:
    - The preloader is visible and spinning during request execution.
    - All project links and buttons are functioning.
    - All forms are validated on the client side.
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the `/signup` route, provided that all input fields have been filled in correctly. If the request is successful, a message is displayed.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a log out button, should appear in its place.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the `/articles` route of the Practicum API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the following: the keyword by which the card was found and the trash can icon for deleting the article.
- [x] <!--2.16--> No errors occur when building or running the project.
- [x] <!--2.16--> The global state variable `currentUser` has been created.
- [x] <!--2.16--> Local storage: the interaction with the JWT token, the results of the search request, and the state variable are set up in the right order.
- Registration and authorization:
    - [x] <!--2.16--> The `/saved-news` route is protected using the `ProtectedRoute` HOC component.
    - [x] <!--2.16--> When trying to access the `/saved-news` route, unauthorized users are redirected to `/signin` with an open authorization popup window.
    - [x] <!--2.16--> When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - [x] <!--2.16--> After successful authorization, the popup window is closed. A link to "Saved articles" and a logout button with the username appear in the header.
    - [x] <!--2.16--> The `/` route is not protected.
    - [x] <!--2.16--> After a successful `onSignOut()` handler call, the user is redirected to `/`.
    - [x] <!--2.16--> The `useHistory()` hook is used correctly.
    - [x] <!--2.16--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- Components:
    - [x] <!--2.16--> Hooks are not used inside conditional statements or loops.
    - [x] <!--2.16--> For class components, effects are described inside the component lifecycle methods.
    - [x] <!--2.16--> Hooks are called in a component's main function.
    - [x] <!--2.16--> The `Main`, `NavBar`, and `SavedNewsHeader` components are subscribed to the `CurrentUserContext` context.
    - [x] <!--2.16--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
    - [x] <!--2.16--> The `currentUser` state variable is created in the `App` root component. This variable is used as a value of the context provider.
    - [x] <!--2.16--> The necessary handlers are defined inside the `App` root component: `onRegister()`, `onLogin()`, and `onSignOut()`. These handlers are passed to the appropriate components: `Register.js`, `Login.js`, and `NavBar.js`.
    - [x] <!--2.16--> The popup components only contain the submit event handlers. Other handlers, such as `handleUpdateUser()`, are described inside the `App` component.
- [x] <!--2.16--> Asynchronous API requests:
	- Requests can be made through the Fetch API or by using XMLHttpRequest.
	- Third-party libraries (such as axios or jQuery) are not used.
	- API requests are contained in separate files: `MainApi.js` and `NewsApi.js`.
	- The chain for processing promises ends with a `catch()` block.
	- The first `then()` handler returns `res.json`.
- [x] <!--2.16--> Naming:
	- Function and variable names use camelCase.
	- Only nouns are used as variable names.
	- Plural nouns are used for `NodeList`.
	- Classes are named using nouns that begin with a capital letter.
	- The `const` keyword is used for variables that won't be changed directly.
	- Class names correspond with their content.
	- Names must not include inappropriate or unclear abbreviations.
- [x] <!--2.16--> No third-party JavaScript libraries are used.
- [x] <!--2.16--> The date parameters of requests to News API must be calculated automatically (7 days prior to the current date). We recommend converting the `from` parameter to a timestamp and vice versa.

### Best Practices

- [x] <!--1.88--> The initial state of state variables contain the correct data type.
- [x] <!--1.88--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--1.88--> API requests are described inside the `App` component.
- [] <!--1.88--> Arrow functions are not passed as handler functions.
- [] <!--1.88--> API error handling:
    - [] <!--1.88--> The user receives a message in case of an error.
    - [x] <!--1.88--> Form fields are blocked while requests are being sent.
- [x] <!--1.88--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.67--> Semantically correct blocks are used for components.
- [x] <!--1.67--> No `<div>` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.67--> The code is clean and easy to understand:
    - The purpose of each listener callback is clear from its name.
    - There are no "magic numbers," i.e. all numeric values are assigned to variables.
    - Сonstants are described in the same scope in which the function is declared. Inside the function, the values of the constants are taken from the closure.

## Number of points: 94

## Additional comments:
- 
```
